### ISSUE TITLE
PPO manual reward functions

### ISSUE THREAD

**Issue by schmidtj3** [huggingface/trl#2363]
In PPO training, I would like to apply customized non-parametric reward function, for instance, rule-based rewards based on textual features of generated texts. In this case, I don't need to use reward model or value model.

For latest trl version (0.12.x), how to achieve this by passing such reward function instead of reward model, into `PPOTrainer` and perform `ppo_trainer.train()`?

Any code with latest trl version can be referenced to?

**Comment by August-murr** [comment]
to get it to work with `ppo_trainer.train`, one idea is to modify the `get_reward` function which is used by `PPOTrainer`.  

Clone the TRL repo and check out the `get_reward` function in `utils.py` 
https://github.com/huggingface/trl/blob/672c96546d9cae7a6d0afba381b189bb3cb2e8b5/trl/trainer/utils.py#L1069-L1093.
Right now, it uses a `torch.nn.Module` to calculate the reward. You can modify it to use your rule-based reward logic instead. Just make sure the function still outputs a `torch.tensor` so `PPOTrainer` doesn’t break.  

You might also need to adjust some references in `ppo_config.py` and `ppo_trainer.py`. For example, remove anything that assumes there’s a reward model being used, since in your case, there won’t be one.
